### PR TITLE
Fix the `str_flatten` function in Redshift

### LINKED ISSUE #804: `str_flatten` does not work in Redshift
The `str_flatten` function is translated to `string_agg` when using the Redshift backend. The `string_agg` function is not supported by Redshift. Instead `str_flatten` should be translated into `LISTAGG`. 

Will submit a pull request soon! :)

Please see the reprex below:

``` r
library(dbplyr)
library(DBI)
library(reprex)

con <- dbConnect(RPostgres::Redshift(), 
                 host = ,
                 dbname = , 
                 port = ,
                 user = , 
                 password = ) 

example_table <- dplyr::tribble(
  ~customer, ~day, ~item,
  "A", 1, "WATER",
  "A", 3, "BREAD",
  "A", 2, "JUICE",
  "B", 1, "APPLE",
  "B", 4, "BANANA",
  "C", 1, "MILK"
)

table_db <- dplyr::copy_to(con, example_table, temporary = T)

table_db %>%
  dplyr::group_by(customer) %>%
  dplyr::summarize(flat_string = str_flatten(item, "-"))
#> Error: Failed to prepare query: ERROR:  function string_agg(character varying, "unknown") does not exist
#> HINT:  No function matches the given name and argument types. You may need to add explicit type casts.

table_db %>%
  dplyr::group_by(customer) %>%
  dbplyr::window_order(day) %>%
  dplyr::mutate(flat_string = str_flatten(item, "-")) 
#> Error: Failed to prepare query: ERROR:  function string_agg(character varying, "unknown") does not exist
#> HINT:  No function matches the given name and argument types. You may need to add explicit type casts.
```

<sup>Created on 2022-03-25 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

### PR DESCRIPTION
**Pull request by hdplsa**
Fixes #804

This pull request adds a Redshift specific translation to `str_flatten`. Currently the default Postgres translation (`string_agg`) is used, but it is not supported in Redshift. The equivalent function in Redshift is `LISTAGG` but it has a slightly different syntax (requires `WITHIN GROUP (ORDER BY ...)` syntax for ordering).

Please check below the reprex with the expected output:

``` r
library(dbplyr)
library(DBI)
library(reprex)

con <- dbConnect(RPostgres::Redshift(), 
                 host = ,
                 dbname = , 
                 port = ,
                 user = , 
                 password = ) 

example_table <- dplyr::tribble(
  ~customer, ~day, ~item,
  "A", 1, "WATER",
  "A", 3, "BREAD",
  "A", 2, "JUICE",
  "B", 1, "APPLE",
  "B", 4, "BANANA",
  "C", 1, "MILK"
)

table_db <- dplyr::copy_to(con, example_table, temporary = T)

table_db %>%
  dplyr::group_by(customer) %>%
  dplyr::summarize(flat_string = str_flatten(item, "-"))
#> # Source:   [?? x 2]
#> # Database: postgres
#> #   []
#>   customer flat_string      
#>   <chr>    <chr>            
#> 1 B        APPLE-BANANA     
#> 2 C        MILK             
#> 3 A        WATER-BREAD-JUICE

table_db %>%
  dplyr::group_by(customer) %>%
  dplyr::summarize(flat_string = str_flatten(item, "-")) %>% 
  dplyr::show_query()
#> <SQL>
#> SELECT "customer", LISTAGG("item", '-') AS "flat_string"
#> FROM "example_table"
#> GROUP BY "customer"

table_db %>%
  dplyr::group_by(customer) %>%
  dbplyr::window_order(day) %>%
  dplyr::mutate(flat_string = str_flatten(item, "-")) 
#> # Source:     [?? x 4]
#> # Database:   postgres
#> #   []
#> # Groups:     customer
#> # Ordered by: day
#>   customer   day item   flat_string      
#>   <chr>    <dbl> <chr>  <chr>            
#> 1 A            1 WATER  WATER-JUICE-BREAD
#> 2 A            2 JUICE  WATER-JUICE-BREAD
#> 3 A            3 BREAD  WATER-JUICE-BREAD
#> 4 C            1 MILK   MILK             
#> 5 B            1 APPLE  APPLE-BANANA     
#> 6 B            4 BANANA APPLE-BANANA

table_db %>%
  dplyr::group_by(customer) %>%
  dbplyr::window_order(day) %>%
  dplyr::mutate(flat_string = str_flatten(item, "-")) %>%
  dplyr::show_query()
#> <SQL
#> SELECT
#>   "customer",
#>   "day",
#>   "item",
#>   LISTAGG("item", '-') WITHIN GROUP (ORDER BY "day") OVER (PARTITION BY "customer") AS "flat_string"
#> FROM "example_table"
```

<sup>Created on 2022-03-25 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

### CODE REVIEW

File: R/backend-redshift.R [hunk 1]
```diff
@@ -76,6 +81,23 @@ sql_translation.RedshiftConnection <- function(con) {
           win_current_frame()
         )
       },
+      # https://docs.aws.amazon.com/redshift/latest/dg/r_LISTAGG.html
+      str_flatten = function(x, collapse) {
+        order <- win_current_order()
+        if(length(order) > 0){
+          sql <- build_sql(sql_expr(LISTAGG(!!x, !!collapse)),
+                           " WITHIN GROUP (ORDER BY ", order, ")")
```
Comment (hadley) [inline R/backend-redshift.R:89]: Are there other functions in redshift that use this syntax? I'm surprised that (say) windowed `mean()` doesn't need the same syntax? Or is `LISTAGG()` the same sort of function as `PERCENTILE_DISC()`?

**Review by hadley — APPROVED** [review]
Can you please add a bullet to the top of `NEWS.md`? It should briefly [describe the change](http://style.tidyverse.org/news.html) and end with `(@yourname, #issuenumber)`.

File: R/backend-redshift.R (hunk 1 above)
Comment (hdplsa) [reply R/backend-redshift.R:89]: From searching the documentation only the `LISTAGG()`, `PERCENTILE_DISC()`, `PERCENTILE_CONT()`, and `ST_COLLECT()`  functions in redshift require the `WITHIN GROUP` clause for the ordering.  The remaining windowed functions use the more general `OVER (PARTITION BY ... ORDER BY ...)` clause. 

Unfortunately, I cannot explain why `LISTAGG()` in particular uses the `WITHIN GROUP` syntax. But this syntax seems to be shared by other database types (e.g.  [Oracle](https://docs.oracle.com/cd/E11882_01/server.112/e41084/functions089.htm#SQLRF30030)).

**Comment by hdplsa** [comment]
Just added the bullet to the NEWS.md. Thanks @hadley.

File: R/backend-redshift.R (hunk 1 above)
Comment (hadley) [reply R/backend-redshift.R:89]: Thanks for the investigation! Unfortunately there seem to be few satisfying explanation for why SQL things are the way they are.

**Comment by hadley** [comment]
Thanks!

**Pull request merged**